espei/run_espei.py

```diff
@@ -125,57 +85,102 @@ def main():
         else:
             return 0
 
-    # run ESPEI fitting
-    # create the scheduler if not passed
-    if args.scheduler == 'MPIPool':
-        from emcee.utils import MPIPool
-        # code recommended by emcee: if not master, wait for instructions then exit
-        client = MPIPool()
-        if not client.is_master():
-            logging.warning('MPIPool is not master. Waiting for instructions...')
-            client.wait()
-            sys.exit(0)
-        logging.info("Using MPIPool on {} MPI ranks".format(client.size))
-    elif not args.scheduler:
-        from distributed import LocalCluster
-        args.scheduler = LocalCluster(n_workers=int(multiprocessing.cpu_count()/2), threads_per_worker=1, processes=True)
-        client = ImmediateClient(args.scheduler)
-        logging.info("Running with dask scheduler: %s [%s cores]" % (args.scheduler, sum(client.ncores().values())))
-        try:
-            logging.info("bokeh server for dask scheduler at localhost:{}".format(client.scheduler_info()['services']['bokeh']))
-        except KeyError:
-            logging.info("Install bokeh to use the dask bokeh server.")
+    # if we aren't checking datasets, then we will check that the input file exists
+    input_file = args.input
+    if input_file is None:
+        raise ValueError('To run ESPEI, provide an input file with the `--input` option.')
+
+    # continue with setup
+    # load the settings
+    ext = os.path.splitext(input_file)[-1]
+    if ext == '.yml' or ext == '.yaml':
+        with open(input_file) as f:
+            input_settings = yaml.load(f)
+    elif ext == '.json':
+        with open(input_file) as f:
+            input_settings = json.load(f)
     else:
-        raise ValueError('Custom schedulers not supported. Use \'MPIPool\' or accept the default Dask LocalCluster.')
+        raise ValueError('Unknown filetype {} for input file {}. YAML and JSON are supported'.format(ext, input_file))
+    run_settings = get_run_settings(input_settings)
+    system_settings = run_settings['system']
+    output_settings = run_settings['output']
+    generate_parameters_settings = run_settings.get('generate_parameters')
+    mcmc_settings = run_settings.get('mcmc')
```

Comment:
I'm thinking everything roughly after here should be broken out into a separate public function. It should be possible to construct `run_settings` programmatically (without writing an input file to disk), and then call a function which will handle the error checking and constructing the call to `fit()`.

Reply:
`run_settings` comes from the public `get_run_settings` function that takes a dict and validates it against a schema. In `main` we determine whether the input file is a YAML or JSON, then deserialize to dict and call `get_run_settings`.

Is that what you're thinking?

As far as constructing the call to fit, the next step is to factor out the call to `fit` into two functions, one for generating parameters and one for fitting them with mcmc. Then, ideally, we just pass in the `mcmc_settings` dict as a `**kwargs` and everything is handled internally

Reply:
I want to be able to pass my own `run_settings` to the fitting routine without it needing to know anything about an input file. I'm thinking about programmatic generation of runs with, e.g., different excess models, different point densities (and different output filenames). It would be convenient to just be able to write a template run settings file, load it, and then modify a dict inside a `for` loop to kick off all the runs.

Reply:
Yeah I think we are in agreement.

Reply:
Right now you could reasonably enumerate all the keyword arguments you want to pass to `fit` as a dict. The refactor I proposed would just make it more simple to do that. Ultimately, `run_espei.py` should just be a way to run ESPEI from the command line with an input file, but you should in principle be able to do without it if you want to work from a custom script, interactively, with a notebook, or whatever.

For the purposes of this PR, we are just switching from a command line argument interface to a file interface to the fit command. What you've brought up is worth opening an issue for, but since this doesn't introduce any regressions on that front, I think it's out of scope here.